flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/OnDataSet.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.scala.extensions.acceptPartialFunctions
+
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.scala.{GroupedDataSet, DataSet}
+
+import scala.reflect.ClassTag
+
+class OnDataSet[T: TypeInformation](ds: DataSet[T]) {
+
+  /**
+    * Applies a function `fun` to each item of the data set
+    *
+    * @param fun The function to be applied to each item
+    * @tparam R The type of the items in the returned data set
+    * @return A dataset of R
+    */
+  def mapWith[R: TypeInformation: ClassTag](fun: T => R): DataSet[R] =
+    ds.map(fun)
+
+  /**
+    * Applies a function `fun` to a partition as a whole
+    *
+    * @param fun The function to be applied on the whole partition
+    * @tparam R The type of the items in the returned data set
+    * @return A dataset of R
+    */
+  def mapPartitionWith[R: TypeInformation: ClassTag](fun: Seq[T] => R): DataSet[R] =
+    ds.mapPartition {
+      (it, out) =>
+        out.collect(fun(it.to[Seq]))
```

Comment:
I think that `it.to[Seq]` can be problematic, since it can happen that an underlying `Vector` is returned here. This means that the whole iterator will be materialized. It is better imo to define `fun: Iterator[T] => R`.